OBSDOCS-1035/TRACING-4077: Write documentation for the k8scluterreceiver component (#75604)
openshift/openshift-docs: modules/otel-collector-components.adoc

Wildcards with `+*+` are accepted under the `cors_allowed_origins`.
To match any origin, enter only `+*+`.

[id="k8scluster-receiver_{context}"]
=== Kubernetes Cluster Receiver

The Kubernetes Cluster Receiver gathers cluster metrics and entity events from the Kubernetes API server. It uses the Kubernetes API to receive information about updates. Authentication for this receiver is only supported through service accounts.

> .OpenTelemetry Collector custom resource with the enabled Kubernetes Cluster Receiver
> [source,yaml]
> ----
>   receivers:
>     k8s_cluster:

Reviewer: Is this block a snippet of a full CR? If so, see repo guidelines on [truncating a YAML file](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#yaml-formatting), e.g.:
```suggestion
.OpenTelemetry Collector custom resource with the enabled Kubernetes Cluster Receiver
[source,yaml]
----
# ...
  receivers:
    k8s_cluster:
```

Author: Thank you. Applying.

Author: I also created a Jira ticket to review the rest of the docs globally.

      distribution: openshift
      collection_interval: 10s
  exporters:
    debug:
  service:
    pipelines:
      metrics:
        receivers: [k8s_cluster]
        exporters: [debug]
      logs/entity_events:
        receivers: [k8s_cluster]
        exporters: [debug]
----

> This receiver requires a set up ServiceAccount, RBAC rules for the ClusterRole, and the ClusterRoleBinding that binds the RBAC with the ServiceAccount.

Reviewer: ISG says "setup" when an adjective. Altho maybe "configured" would feel a little more natural here? 

Also marking up the [objects](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#object-references):
```suggestion
This receiver requires a configured `ServiceAccount` object, RBAC rules for the `ClusterRole` object, and the `ClusterRoleBinding` object that binds the RBAC with the `ServiceAccount`.
```
[Alternatively](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#general-references):
```suggestion
This receiver requires a configured service account, RBAC rules for the cluster role, and the cluster role binding that binds the RBAC with the service account.
```

Author: :slightly_smiling_face: I used `set up` as a participle.

> .The ServiceAccount

Reviewer: Suggest avoiding leading off with "The" if possible.
```suggestion
.`ServiceAccount` object
```

Author: Thank you. Applied.

[source,yaml]
----
apiVersion: v1
kind: ServiceAccount
metadata:
  labels:
    app: otelcontribcol
  name: otelcontribcol
----

> .The RBAC rules for the ClusterRole

Reviewer: ```suggestion
.RBAC rules for the `ClusterRole` object
```

Author: Thank you. Applying.

[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1

> kind: ClusterRole

Reviewer: Consider also the [collapsible option](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#general-references) for this longest block, or even just all 3 for consistency.

Author: Thank you for the suggestion.
I like using the collapsible option where possible. However, I'll hold off adding complexity to the docs until migration to the new toolchain.

metadata:
  name: otelcontribcol
  labels:
    app: otelcontribcol
rules:
- apiGroups:
  - quota.openshift.io
  resources:
  - clusterresourcequotas
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - ""
  resources:
  - events
  - namespaces
  - namespaces/status
  - nodes
  - nodes/spec
  - pods
  - pods/status
  - replicationcontrollers
  - replicationcontrollers/status
  - resourcequotas
  - services
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - apps
  resources:
  - daemonsets
  - deployments
  - replicasets
  - statefulsets
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - extensions
  resources:
  - daemonsets
  - deployments
  - replicasets
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - batch
  resources:
  - jobs
  - cronjobs
  verbs:
  - get
  - list
  - watch
- apiGroups:
    - autoscaling
  resources:
    - horizontalpodautoscalers
  verbs:
    - get
    - list
    - watch
----

> .The ClusterRoleBinding

Reviewer: ```suggestion
.`ClusterRoleBinding` object
```

Author: Thank you. Applied.

    app: otelcontribcol
rules:
- apiGroups:
  - ""
  resources:
  - events
  - namespaces
  - namespaces/status
  - nodes
  - nodes/spec
  - pods
  - pods/status
  - replicationcontrollers
  - replicationcontrollers/status
  - resourcequotas
  - services
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - apps
  resources:
  - daemonsets
  - deployments
  - replicasets
  - statefulsets
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - extensions
  resources:
  - daemonsets
  - deployments
  - replicasets
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - batch
  resources:
  - jobs
  - cronjobs
  verbs:
  - get
  - list
  - watch
- apiGroups:
    - autoscaling
  resources:
    - horizontalpodautoscalers
  verbs:
    - get
    - list
    - watch
----

And the ClusterRoleBinding to bind the RBAC with the ServiceAccount
.Kubernetes ClusterRoleBinding resource
[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: otelcontribcol
  labels:
    app: otelcontribcol
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: otelcontribcol
subjects:
- kind: ServiceAccount
  name: otelcontribcol
  namespace: default
----

> You can enable OpenShift support to collect OpenShift specific metrics in addition to the default kubernetes ones. To do this,set the distribution key to openshift.

Author: Updated.

.OpenTelemetry Collector custom resource with the enabled Kubernetes cluster receiver with openshift distribution
[source,yaml]
----
  k8s_cluster:
    distribution: openshift
----

> Also if you enable OpenShift distribution you need to add the following rules to your ClusterRole.

Author: @rubenvp8510, can I incorporate these lines into this ClusterRole: https://github.com/openshift/openshift-docs/pull/75604/files#diff-294ec585891d4f53f51203c945436edac35259850eb19f99f502661759203d21R242? Or does this require another ClusterRole?

Author: :slightly_smiling_face: I mean the downstream OTEL product is only supported on OpenShift, so we have to make the OpenShift configuration the default settings rather than optional.

Reviewer 2: It will be the same cluster role.

Author: Thank you, Ruben. I'm incorporating it then.
